src/main/kotlin/com/exactpro/th2/common/utils/message/transport/MessageUtils.kt

```diff
@@ -0,0 +1,126 @@
+    addMessage(this@toGroup)
+}.build()
+
+fun MessageId.toProto(book: String, sessionGroup: String): MessageID = MessageID.newBuilder().also {
+    it.bookName = book
+    it.direction = if (direction == INCOMING) FIRST else SECOND
+    it.sequence = sequence
+    it.timestamp = timestamp.toTimestamp()
+
+    it.addAllSubsequence(subsequence)
+
+    it.connectionIdBuilder.also { connectionId ->
+        connectionId.sessionGroup = sessionGroup.ifBlank { sessionAlias }
+        connectionId.sessionAlias = sessionAlias
+    }
+}.build()
+
+fun MessageId.toProto(groupBatch: GroupBatch): MessageID = toProto(groupBatch.book, groupBatch.sessionGroup)
+
+fun ParsedMessage.toProto(book: String, sessionGroup: String): ProtoMessage = ProtoMessage.newBuilder().apply {
+    metadataBuilder.apply {
+        id = this@toProto.id.toProto(book, sessionGroup)
+        messageType = this@toProto.type
+        protocol = this@toProto.protocol
+        putAllProperties(this@toProto.metadata)
+    }
+    body.forEach { (key, value) -> addField(key, value.toValue()) }
+    eventId?.let { parentEventId = it.toProto() }
+}.build()
+
+fun ParsedMessage.toTreeTable(): TreeTable = TreeTableBuilder().apply {
+    for ((key, value) in body) {
+        row(key, value.toTreeTableEntry())
+    }
+}.build()
+
+private fun Any?.toTreeTableEntry(): TreeTableEntry {
+    if (this == null) {
+        return RowBuilder()
+            .column(MessageTableColumn("null"))
+            .build()
+    }
+    return when (this) {
+        is Map<*, *> -> {
+            CollectionBuilder().apply {
+                forEach { (key, value) ->
+                    row(key.toString(), value.toTreeTableEntry())
+                }
+            }.build()
+        }
+
+        is List<*> -> {
+            CollectionBuilder().apply {
+                forEachIndexed { index, nestedValue ->
+                    row(index.toString(), nestedValue.toTreeTableEntry())
+                }
+            }.build()
+        }
+
+        is String -> RowBuilder().column(MessageTableColumn(this)).build()
+        is Number -> RowBuilder().column(MessageTableColumn(convertToString())).build()
+        else -> error("Unsupported ${this::class.simpleName} number type, value $this")
+    }
+}
+
+fun Number.convertToString(): String = when (this) {
+    is Byte,
+    is Short,
+    is Int -> toString()
+
+    is Float,
+    is Double -> BigDecimal(toString()).stripTrailingZeros().toPlainString()
+
+    is BigDecimal -> stripTrailingZeros().toPlainString()
+    else -> error("Unsupported ${this::class.simpleName} number type, value $this")
+}
+
+@Suppress("UNCHECKED_CAST")
+fun <T : Message.Builder<T>> Message.Builder<T>.addMetadataProperty(key: String, value: String): T {
+    metadataBuilder().put(key, value)
+    return this as T
+}
+
+fun ParsedMessage.FromMapBuilder.addField(name: String, value: Any?): ParsedMessage.FromMapBuilder = apply {
+    bodyBuilder().put(name, value)
+}
```

Comment:
This is not very useful when you work with a builder in Java... I believe that should be kept as a part of the interface itself

Reply:
Reverted